### PR TITLE
CMake fix, refactoring and remove deprecated files

### PR DESCRIPTION
**Pull request by NIKEA-SOFT**
List of changes:
- Refactoring code
- Clang-format code
- Fix CMake flags for multi-configurations
- Fix memory leaks
- Fix undefined behaviour
- Remove unused or deprecated files after refactoring

Also for CMake: VisualStudio now displays most of the header files in the project

### CODE REVIEW

File: ai/default/aisettler.cpp [hunk 1]
```diff
@@ -377,7 +377,7 @@ static struct cityresult *cityresult_fill(struct ai_type *ait,
     }
 
     if (result->tdc_hash->contains(cindex)) {
-      NFCPP_FREE(result->tdc_hash->value(cindex));
```
Comment (pranavtifr) [inline ai/default/aisettler.cpp:380]: Why were these macros removed ? They do the exact same thing as the delete.

Comment (lmoureaux) [reply ai/default/aisettler.cpp:380]: Readability

Comment (pranavtifr) [reply ai/default/aisettler.cpp:380]: So, The aim is to remove all the macros ? Someone should have told me that when I was fixing the malloc-dealloc mismatches lol

**Comment by lmoureaux** [comment]
Codacy: it failed to detect issues that moved to another code location. Feel free to ignore everything.

File: ai/default/aisettler.cpp (hunk 1 above)
Comment (lmoureaux) [reply ai/default/aisettler.cpp:380]: Over time yes. I don't even understand what the various prefixes are for, need to look at the definition every time :D

**Comment by lmoureaux** [comment]
So it looks like this branch has the same rebasing problem as the previous one. Specifically, it was rebased on top of `f31e3589dd2e4f48649ae619d70f227e0554c799`, whereas current `master` is at `0926f8bc208cc552cc6a16dd8cd7e90cc4ee85f7`.

File: client/gui-qt/hudwidget.cpp [hunk 2]
```diff
@@ -828,11 +828,10 @@ hud_action::hud_action(QWidget *parent, const QIcon &icon,
  */
 void hud_action::paintEvent(QPaintEvent *event)
 {
-  QRect rx, ry, rz;
+  ;
```
Comment (lmoureaux) [inline client/gui-qt/hudwidget.cpp:831]: ??

**Review by lmoureaux — CHANGES_REQUESTED** [review]
Had a first, cursory look at the code. General comments:

* Can you prove that for every `NFCPP_FREE` your removed, the freed pointer cannot be null? The macro checks it.
* Do not introduce commented code. If it's not needed, remove it.
* Avoid reindenting comments. `clang-format` doesn't do it on its own, so it must be you.
* `Q_UNUSED` is fine.

File: client/gui-qt/hudwidget.cpp (hunk 2 above)
Comment (NIKEA-SOFT) [reply client/gui-qt/hudwidget.cpp:831]: > ??

Unused variable `ry`

**Comment by NIKEA-SOFT** [comment]
> Had a first, cursory look at the code. General comments:
> 
> * Can you prove that for every `NFCPP_FREE` your removed, the freed pointer cannot be null? The macro checks it.
> * Do not introduce commented code. If it's not needed, remove it.
> * Avoid reindenting comments. `clang-format` doesn't do it on its own, so it must be you.
> * `Q_UNUSED` is fine.

1) At least I tried to stick to the same behavior as macros. At the same time, I dropped the checks for nullptr since they are redundant. Not all macros assign nullptr after calling delete, so I tried to follow the same behavior.
2) Understood
3) Formatting comments is related to calling `clang-format`, not `git clang-format'. My previous changes could not pass the git action check due to the fact that `git clang-format` for some reason did not want to format comments and the action failed during the check. This problem was solved when I started using `clang-format` locally, without git, but this led to the fact that it formatted a lot of comments and most of the changes are related to indentation.
4) Why should we use Q_UNUSED when we can use attributes supported by C++17 from the standard?

**Comment by lmoureaux** [comment]
> At least I tried to stick to the same behavior as macros. At the same time, I dropped the checks for nullptr since they are redundant. Not all macros assign nullptr after calling delete, so I tried to follow the same behavior.

This is guaranteed to crash:
```C++
int *x = nullptr;
delete x;
```
So I'm failing to see how the null check is redundant...

> Why should we use Q_UNUSED when we can use attributes supported by C++17 from the standard?

It's idiomatic Qt and has been around for decades, so people used to Qt know about it. `[[maybe_unused]]` is just 5 years old, and C++17 is only reaching the industry now... I'm not saying we shouldn't use it, I'm fine with either syntax. Just don't change for the sake of changing. Also a personal opinion: attributes on variables look ugly and make it harder to parse the declaration.

File: client/gui-qt/hudwidget.cpp (hunk 2 above)
Comment (lmoureaux) [reply client/gui-qt/hudwidget.cpp:831]: It seems you introduced a line with just a `;`...

**Comment by NIKEA-SOFT** [comment]
> > At least I tried to stick to the same behavior as macros. At the same time, I dropped the checks for nullptr since they are redundant. Not all macros assign nullptr after calling delete, so I tried to follow the same behavior.
> 
> This is guaranteed to crash:
> 
> ```c++
> int *x = nullptr;
> delete x;
> ```
> 
> So I'm failing to see how the null check is redundant...
> 
> > Why should we use Q_UNUSED when we can use attributes supported by C++17 from the standard?
> 
> It's idiomatic Qt and has been around for decades, so people used to Qt know about it. `[[maybe_unused]]` is just 5 years old, and C++17 is only reaching the industry now... I'm not saying we shouldn't use it, I'm fine with either syntax. Just don't change for the sake of changing. Also a personal opinion: attributes on variables look ugly and make it harder to parse the declaration.

As for `nullptr` and `delete`, you are mistaken, the standard guarantees that if `delete` was not overloaded by the user, that is, it is the default delete operator, then calling delete with passing a pointer that is nullptr has no effect.

These are not changes for the sake of changes, these are the capabilities of the standard, I would prefer them than using macros. 
I also don't understand why we should look at how others do it, because we are not talking about other, we are talking about freeciv21. What is the point of porting freeciv to C++ if we abandon the new features of the standard in order to please those who are behind the times?
I'm not trying to make as many changes as possible, I just don't like reading the code that's there, and I want to make it more readable and modern.
If I come across some part of the code that I can make more readable, then I try to do it. It doesn't take much time, but it allows you to make the code more readable little by little.

**Comment by lmoureaux** [comment]
> As for nullptr and delete, you are mistaken, the standard guarantees that if delete was not overloaded by the user, that is, it is the default delete operator, then calling delete with passing a pointer that is nullptr has no effect.

Wow I didn't know this. I've been spoiled by C too much.

>  I also don't understand why we should look at how others do it, because we are not talking about other, we are talking about freeciv21. What is the point of porting freeciv to C++ if we abandon the new features of the standard in order to please those who are behind the times?

Well... `Q_UNUSED` and `[[maybe_unused]]` are semantically identical. The existence of a new way doesn't mean that the old way is invalid.

**Comment by NIKEA-SOFT** [comment]
> > As for nullptr and delete, you are mistaken, the standard guarantees that if delete was not overloaded by the user, that is, it is the default delete operator, then calling delete with passing a pointer that is nullptr has no effect.
> 
> Wow I didn't know this. I've been spoiled by C too much.

C standard guarantees the same.
Most likely you meant a double delete/free
```
int* ptr = new int{};
delete ptr;
delete ptr; // UB
```

```
int* ptr = new int{};
delete ptr;
ptr = nullptr;
delete ptr; // no effect
```

File: ai/classic/classicai.cpp [hunk 3]
```diff
@@ -619,11 +619,11 @@ bool fc_ai_classic_setup(struct ai_type *ai)
...
-    ai->funcs.city_created = nullptr;
-    ai->funcs.city_destroyed = nullptr;
-    ai->funcs.city_got = nullptr;
-    ai->funcs.city_lost = nullptr;
-  */
+  ai->funcs.city_created = nullptr;
+  ai->funcs.city_destroyed = nullptr;
+  ai->funcs.city_got = nullptr;
+  ai->funcs.city_lost = nullptr;
+*/
```
Comment (lmoureaux) [inline ai/classic/classicai.cpp:626]: Remove commented code.

File: ai/classic/classicai.cpp [hunk 4]
```diff
@@ -637,18 +637,18 @@ bool fc_ai_classic_setup(struct ai_type *ai)
...
 
   /* FIXME: We should allocate memory only for units owned by
-     default ai in unit_got. We track no data
-     about enemy units.
-     But advisors code still depends on some default ai data (role) to be
-     always allocated. */
+   default ai in unit_got. We track no data
+   about enemy units.
+   But advisors code still depends on some default ai data (role) to be
+   always allocated. */
```
Comment (lmoureaux) [inline ai/classic/classicai.cpp:643]: Don't reindent comments

File: ai/classic/classicai.cpp [hunk 5]
```diff
@@ -637,18 +637,18 @@ bool fc_ai_classic_setup(struct ai_type *ai)
...
-    ai->funcs.unit_got = dai_unit_init;
-    ai->funcs.unit_lost = dai_unit_close;
-  */
+  ai->funcs.unit_alloc = nullptr;
+  ai->funcs.unit_free = nullptr;
+  ai->funcs.unit_created = nullptr;
+  ai->funcs.unit_destroyed = nullptr;
+  ai->funcs.unit_got = dai_unit_init;
+  ai->funcs.unit_lost = dai_unit_close;
+*/
```
Comment (lmoureaux) [inline ai/classic/classicai.cpp:651]: Don't reindent comments

File: ai/default/aidata.cpp [hunk 6]
```diff
@@ -147,10 +147,10 @@ void dai_data_phase_begin(struct ai_type *ait, struct player *pplayer,
...
 
   /* Note that this refreshes advisor data if needed. ai_plr_data_get()
-     is expected to refresh advisor data if needed, and ai_plr_data_get()
-     depends on this call
-     ai_plr_data_get()->ai_data_phase_begin()->adv_data_get() to do it.
-     If you change this, you may need to adjust ai_plr_data_get() also. */
+   is expected to refresh advisor data if needed, and ai_plr_data_get()
+   depends on this call
+   ai_plr_data_get()->ai_data_phase_begin()->adv_data_get() to do it.
+   If you change this, you may need to adjust ai_plr_data_get() also. */
```
Comment (lmoureaux) [inline ai/default/aidata.cpp:153]: Don't reindent comments

File: ai/default/aidata.cpp [hunk 7]
```diff
@@ -162,7 +162,7 @@ void dai_data_phase_begin(struct ai_type *ait, struct player *pplayer,
   adv = adv_data_get(pplayer, &caller_closes);
 
   /* Store current number of known continents and oceans so we can compare
-     against it later in order to see if ai data needs refreshing. */
+   against it later in order to see if ai data needs refreshing. */
```
Comment (lmoureaux) [inline ai/default/aidata.cpp:165]: Don't reindent comments

File: ai/default/aidata.cpp [hunk 8]
```diff
@@ -314,7 +314,7 @@ struct ai_plr *dai_plr_data_get(struct ai_type *ait, struct player *pplayer,
   fc_assert_ret_val(ai != nullptr, nullptr);
 
   /* This assert really is required. See longer comment
-     in adv_data_get() for equivalent code. */
+   in adv_data_get() for equivalent code. */
```
Comment (lmoureaux) [inline ai/default/aidata.cpp:317]: Don't reindent comments

File: ai/default/aidiplomat.cpp [hunk 9]
```diff
@@ -119,9 +119,9 @@ void dai_choose_diplomat_defensive(struct ai_type *ait,
                                    struct adv_choice *choice, int def)
 {
   /* Build a diplomat if our city is threatened by enemy diplomats, and
-     we have other defensive troops, and we don't already have a diplomat
-     to protect us. If we see an enemy diplomat and we don't have diplomat
-     tech... race it! */
+   we have other defensive troops, and we don't already have a diplomat
+   to protect us. If we see an enemy diplomat and we don't have diplomat
+   tech... race it! */
```
Comment (lmoureaux) [inline ai/default/aidiplomat.cpp:124]: Don't reindent comments

File: ai/default/aidiplomat.cpp [hunk 10]
```diff
@@ -216,7 +216,7 @@ void dai_choose_diplomat_offensive(struct ai_type *ait,
         && (incite_cost < INCITE_IMPOSSIBLE_COST)
         && (incite_cost < pplayer->economic.gold - expenses)) {
       /* incite gain (FIXME: we should count wonders too but need to
-         cache that somehow to avoid CPU hog -- Per) */
+   cache that somehow to avoid CPU hog -- Per) */
```
Comment (lmoureaux) [inline ai/default/aidiplomat.cpp:219]: Don't reindent comments

File: ai/default/aiferry.cpp [hunk 11]
```diff
@@ -91,12 +91,12 @@
...
-     data structure, not reading them, so we have no need for extra
-     arrangements dai_plr_data_get() would do compared to
-     def_ai_player_data()
-  */
+   We are only initializing player data structures and dai_plr_data_get()
+   would try to use it uninitialized. We are only setting values to
+   data structure, not reading them, so we have no need for extra
+   arrangements dai_plr_data_get() would do compared to
+   def_ai_player_data()
+*/
```
Comment (lmoureaux) [inline ai/default/aiferry.cpp:99]: Don't reindent comments

File: ai/default/aitools.cpp [hunk 12]
```diff
@@ -750,7 +750,7 @@ void dai_unit_new_task(struct ai_type *ait, struct unit *punit,
   }
 
   /* Map ai tasks to advisor tasks. For most ai tasks there is
-     no advisor, so AUT_NONE is set. */
+   no advisor, so AUT_NONE is set. */
```
Comment (lmoureaux) [inline ai/default/aitools.cpp:753]: Don't reindent comments

File: ai/default/aitools.cpp [hunk 13]
```diff
@@ -778,8 +778,8 @@ bool dai_unit_make_homecity(struct unit *punit, struct city *pcity)
      * so it would be stupid to give it one. There can also be good reasons
      * why it doesn't have a homecity. */
     /* However, until we can do something more useful with them, we
-       will assign explorers to a city so that they can be disbanded for
-       the greater good -- Per */
+   will assign explorers to a city so that they can be disbanded for
+   the greater good -- Per */
```
Comment (lmoureaux) [inline ai/default/aitools.cpp:782]: Don't reindent comments

File: ai/default/aiunit.cpp [hunk 14]
```diff
@@ -1879,7 +1879,7 @@ static void dai_military_attack(struct ai_type *ait, struct player *pplayer,
     };
   } else {
     /* You can still have some moves left here, but barbarians should
-       not sit helplessly, but advance towards nearest known enemy city */
+   not sit helplessly, but advance towards nearest known enemy city */
```
Comment (lmoureaux) [inline ai/default/aiunit.cpp:1882]: Don't reindent comments. I think you get the idea, not commenting on the other ones...

File: client/gui-qt/dialogs.cpp [hunk 15]
```diff
@@ -3779,18 +3749,15 @@ void popup_upgrade_dialog(struct unit_list *punits)
...
-    for (int id : not_ptr) {
-      punit = game_unit_by_number(id);
-      if (punit) {
-        request_unit_upgrade(punit);
-      }
-    }
-  });
+  QObject::connect(ask, &hud_message_box::accepted,
+                   [unit_ids = std::move(unit_ids)] {
+                     for (auto id : unit_ids) {
```
Comment (lmoureaux) [inline client/gui-qt/dialogs.cpp:3754]: you may want a `qAsConst` here

File: client/gui-qt/qtg_cxxside.h [hunk 16]
```diff
@@ -18,24 +18,24 @@
 
 class QTcpSocket;
 
-void setup_gui_funcs();
+// void setup_gui_funcs();
```
Comment (lmoureaux) [inline client/gui-qt/qtg_cxxside.h:21]: Remove

File: client/gui-qt/qtg_cxxside.h [hunk 17]
```diff
@@ -62,7 +61,7 @@ bool qtg_city_dialog_is_open(struct city *pcity);
 
 bool qtg_request_transport(struct unit *pcargo, struct tile *ptile);
 
-void qtg_update_infra_dialog();
+// void qtg_update_infra_dialog();
```
Comment (lmoureaux) [inline client/gui-qt/qtg_cxxside.h:64]: Remove

File: client/include/citydlg_g.h [hunk 18]
```diff
@@ -10,20 +10,17 @@
...
-GUI_FUNC_PROTO(void, popdown_all_city_dialogs, void)
-GUI_FUNC_PROTO(void, real_city_dialog_refresh, struct city *pcity)
-GUI_FUNC_PROTO(void, refresh_unit_city_dialogs, struct unit *punit)
-GUI_FUNC_PROTO(bool, city_dialog_is_open, struct city *pcity)
+// void real_city_dialog_popup(city *pcity);
+// void popdown_city_dialog(city* pcity);
+// void popdown_all_city_dialogs();
+// void real_city_dialog_refresh(city* pcity);
+// void refresh_unit_city_dialogs(unit* punit);
+// bool city_dialog_is_open(city* pcity);
```
Comment (lmoureaux) [inline client/include/citydlg_g.h:21]: Remove

File: client/include/dialogs_g.h [hunk 19]
```diff
@@ -23,69 +23,54 @@
...
-GUI_FUNC_PROTO(void, popdown_all_game_dialogs, void)
-
-GUI_FUNC_PROTO(bool, request_transport, struct unit *pcargo,
-               struct tile *ptile)
+void unit_select_dialog_update_real(void *unused);
+void races_toggles_set_sensitive();
+void races_update_pickable(bool nationset_change);
+// void popup_combat_info(int attacker_unit_id, int defender_unit_id,
+//                int attacker_hp, int defender_hp,
+//                bool make_att_veteran, bool make_def_veteran);
```
Comment (lmoureaux) [inline client/include/dialogs_g.h:42]: Remove

File: client/include/dialogs_g.h [hunk 20]
```diff
@@ -23,69 +23,54 @@
...
+void popup_sabotage_dialog(unit *actor, city *pcity, const action *paction);
+void popup_pillage_dialog(unit *punit, bv_extras extras);
+void popup_upgrade_dialog(unit_list *punits);
+void popup_disband_dialog(unit_list *punits);
+void popup_tileset_suggestion_dialog();
+void popup_soundset_suggestion_dialog();
+void popup_musicset_suggestion_dialog();
+bool popup_theme_suggestion_dialog([[maybe_unused]] const char *theme_name);
+void show_tech_gained_dialog([[maybe_unused]] Tech_type_id tech);
+// bool handmade_scenario_warning();
```
Comment (lmoureaux) [inline client/include/dialogs_g.h:73]: Remove

File: client/include/dialogs_g.h [hunk 21]
```diff
@@ -23,69 +23,54 @@
...
+void popdown_all_game_dialogs();
+// bool request_transport(unit *pcargo, tile *ptile);
+// void update_infra_dialog(); // defined in gui_interface.cpp
```
Comment (lmoureaux) [inline client/include/dialogs_g.h:76]: Remove

File: client/include/gui_main_g.h [hunk 22]
```diff
@@ -19,28 +19,25 @@ class QTcpSocket;
 // common
 #include "fc_types.h"
 
-#include "gui_proto_constructor.h"
-
 void ui_main();
 void ui_exit();
 
-GUI_FUNC_PROTO(void, options_extra_init, void)
+// void options_extra_init();
```
Comment (lmoureaux) [inline client/include/gui_main_g.h:25]: Remove

File: client/include/gui_main_g.h [hunk 23]
```diff
@@ -19,28 +19,25 @@ class QTcpSocket;
...
 
-GUI_FUNC_PROTO(void, real_conn_list_dialog_update, void *)
-GUI_FUNC_PROTO(void, sound_bell, void)
-GUI_FUNC_PROTO(void, add_net_input, QTcpSocket *)
-GUI_FUNC_PROTO(void, remove_net_input, void)
+// void real_conn_list_dialog_update(void*);
+// void sound_bell();
+// void add_net_input(QTcpSocket *)
+// void remove_net_input();
```
Comment (lmoureaux) [inline client/include/gui_main_g.h:30]: Remove

File: client/include/gui_main_g.h [hunk 24]
```diff
@@ -19,28 +19,25 @@ class QTcpSocket;
...
 
-GUI_FUNC_PROTO(void, add_idle_callback, void(callback)(void *), void *data)
+// void add_idle_callback(void(callback)(void *), void *data);
 
-GUI_FUNC_PROTO(void, insert_client_build_info, char *outbuf, size_t outlen)
+void insert_client_build_info(char *outbuf, size_t outlen);
 
-GUI_FUNC_PROTO(void, gui_update_font, const QString &font_name,
-               const QString &font_value)
+// void gui_update_font(const QString &font_name, const QString &font_value);
```
Comment (lmoureaux) [inline client/include/gui_main_g.h:40]: Few more to remove here

File: client/include/mapview_g.h [hunk 25]
```diff
@@ -22,36 +22,30 @@
...
-#include "canvas_g.h"
+void update_info_label();
+void update_unit_info_label(unit_list *punitlist);
+void update_mouse_cursor(cursor_type new_cursor_type);
+// void update_timeout_label();
+void update_turn_done_button(bool do_restore);
+void update_city_descriptions();
 
-#include "gui_proto_constructor.h"
+// void start_turn();
```
Comment (lmoureaux) [inline client/include/mapview_g.h:32]: Remove

File: client/include/pages_g.h [hunk 26]
```diff
@@ -22,13 +22,11 @@
 #define SPECENUM_VALUE5 PAGE_GAME     // In game page.
 #include "specenum_gen.h"
 
-#include "gui_proto_constructor.h"
-
-GUI_FUNC_PROTO(void, real_set_client_page, enum client_pages page)
-GUI_FUNC_PROTO(enum client_pages, get_current_client_page, void)
-GUI_FUNC_PROTO(void, update_start_page, void)
+// void real_set_client_page(client_pages page);
+// client_pages get_current_client_page();
```
Comment (lmoureaux) [inline client/include/pages_g.h:26]: Remove

File: client/include/themes_g.h [hunk 27]
```diff
@@ -10,13 +10,11 @@ received a copy of the GNU General Public License along with Freeciv21.
 ***********************************************************************/
 #pragma once
 
-#include "gui_proto_constructor.h"
-
 class QString;
 class QStringList;
 
-GUI_FUNC_PROTO(void, gui_clear_theme, void)
+// void gui_clear_theme();
```
Comment (lmoureaux) [inline client/include/themes_g.h:16]: Remove

File: client/include/themes_g.h [hunk 28]
```diff
@@ -10,13 +10,11 @@ received a copy of the GNU General Public License along with Freeciv21.
...
 
 void gui_load_theme(const QString &directory, const QString &theme_name);
 QStringList get_useable_themes_in_directory(QString &directory);
-QStringList get_gui_specific_themes_directories(int *count);
+// QStringList get_gui_specific_themes_directories(int *count);
```
Comment (lmoureaux) [inline client/include/themes_g.h:20]: Remove

File: client/packhand.cpp [hunk 29]
```diff
@@ -2364,7 +2365,7 @@ void handle_player_info(const struct packet_player_info *pinfo)
 
   if (pplayer->economic.infra_points != pinfo->infrapoints) {
     pplayer->economic.infra_points = pinfo->infrapoints;
-    update_infra_dialog();
+    // update_infra_dialog(); // this functions is empty lambda
```
Comment (lmoureaux) [inline client/packhand.cpp:2368]: I don't understand the comment, possibly remove

File: common/aicore/cm.h [hunk 30]
```diff
@@ -22,6 +22,21 @@ struct cm_parameter {
 
   int factor[O_LAST];
   int happy_factor;
+
+  cm_parameter &operator=(const cm_parameter &other)
```
Comment (lmoureaux) [inline common/aicore/cm.h:26]: Add a docstring

**Review by lmoureaux — CHANGES_REQUESTED** [review]
Quickly went through the files not "collapsed" by Github with detailed comments. I may have missed things...

**Pull request closed without merging**